### ISSUE TITLE
Switch models on the fly without changing or redeploying code

### ISSUE THREAD

**Issue by ttzeng** [ttzeng/chat-bot#11]
[Large Language Model (LLM)](https://aws.amazon.com/what-is/large-language-model) vendors are evolving their models like [this](https://openai.com/index/gpt-4o-mini-advancing-cost-efficient-intelligence/), and right now the chat bot on the `gpt` branch switches between the models by [comparing the value](https://github.com/ttzeng/chat-bot/blob/gpt/main.py#L88-L96) associated with the key `model` in the redis data store, the actual model used still rely on the default parameter passed to each API. As a result, code changes and re-deployment are required to evaluate a new model, and this should be improved to switch the model on the fly.

**Comment by ttzeng** [comment]
Switch models under the hook by setting the value in the redis data store:
- OpenAI
  `https://us-west1-ttzeng-gcp.cloudfunctions.net/redis-cloud?model={"provider":"openai","model":"gpt-4o-mini"}`
- Gemini
  `https://us-west1-ttzeng-gcp.cloudfunctions.net/redis-cloud?model={"provider":"gemini","model":"gemini-2.0-flash-lite"}`
- Claude
  `https://us-west1-ttzeng-gcp.cloudfunctions.net/redis-cloud?model={"provider":"claude","model":"claude-3-7-sonnet-20250219"}`

**Comment by ttzeng** [comment]
To configure model dependent parameters, use `parameters` as the key, and pass the required parameters in a dictionary object as the value to the key like below:

* OpenAI
```
https://us-west1-ttzeng-gcp.cloudfunctions.net/redis-cloud?model={"provider":"openai","model":"gpt-4o-mini","parameters":{ "temperature":0.8}}
```
* Gemini
```
https://us-west1-ttzeng-gcp.cloudfunctions.net/redis-cloud?model={"provider":"gemini","model":"gemini-2.0-flash-lite","parameters":{}}
```
* Claude
```
https://us-west1-ttzeng-gcp.cloudfunctions.net/redis-cloud?model={"provider":"claude","model":"claude-3-7-sonnet-20250219","parameters":{"max_tokens":1024,"temperature":0.8}}
```

**Comment by ttzeng** [comment]
Google [releases][1] his most intelligent AI model - **Gemini 2.5**, it's a thinking model, capable of reasoning through their thoughts before responding, resulting in enhanced performance and improved accuracy. Let's see how Gemini 2.5 Pro experimental version explained `What are reasoning models`:

---
Okay, let's break down what "reasoning models" are, particularly in the context of Artificial Intelligence (AI).

At its core, a **reasoning model** is an AI system designed to mimic or perform human-like reasoning processes. Instead of just recognizing patterns or making predictions based purely on statistical correlations (like many basic machine learning models), reasoning models aim to:

1.  **Draw Logical Conclusions:** Use given information (premises) to arrive at new, valid conclusions.
2.  **Solve Problems Step-by-Step:** Break down complex problems into manageable steps and follow a logical sequence to find a solution.
3.  **Understand Relationships:** Grasp causal links, spatial relationships, temporal sequences, and other connections between concepts or entities.
4.  **Handle Ambiguity and Incompleteness:** Make plausible inferences or generate hypotheses even when information is missing or uncertain.
5.  **Provide Explanations:** Justify their conclusions or actions based on the reasoning steps taken (explainability).


**Think of the difference:**

*   **Non-Reasoning (Pattern Matching) Model:** Sees 100 pictures of apples and learns to identify an apple. It knows *what* an apple looks like.
*   **Reasoning Model:** Knows "Apples grow on trees," "Trees need sunlight," and "This room has no sunlight." It can infer, "An apple tree probably won't grow well in this room," even without ever seeing that specific scenario. It understands *relationships* and can make *inferences*.

**Types of Reasoning Often Modeled:**

AI reasoning models attempt to replicate various forms of human reasoning, including:

1.  **Deductive Reasoning:** Moving from general rules to specific conclusions (e.g., All humans are mortal. Socrates is human. -> Socrates is mortal).
2.  **Inductive Reasoning:** Generalizing from specific observations to broader rules (e.g., Every swan I've seen is white. -> All swans are probably white). This is often probabilistic.
3.  **Abductive Reasoning:** Finding the most likely explanation for an observation (e.g., The ground is wet. -> It probably rained).
4.  **Analogical Reasoning:** Applying knowledge from one situation to a similar, new situation.
5.  **Causal Reasoning:** Understanding cause-and-effect relationships.
6.  **Spatial Reasoning:** Understanding relationships between objects in space.
7.  **Temporal Reasoning:** Understanding sequences and relationships in time.
8.  **Common Sense Reasoning:** Using the vast background knowledge about the world that humans typically possess.

**Approaches to Building Reasoning Models:**

1.  **Symbolic AI (GOFAI - Good Old-Fashioned AI):**
    *   Uses explicit rules, logic (like predicate calculus), ontologies, and knowledge graphs.
    *   Reasoning is done by manipulating symbols according to pre-defined rules.
    *   **Pros:** Often highly explainable, precise.
    *   **Cons:** Can be brittle (fail on unexpected input), requires extensive knowledge engineering, hard to scale.

2.  **Connectionist / Neural Networks (Especially Large Language Models - LLMs):**
    *   Models like GPT-4, Claude, Gemini, etc., demonstrate significant reasoning capabilities that seem to *emerge* from being trained on massive amounts of text and code.
    *   Techniques like **Chain-of-Thought (CoT) prompting** explicitly guide these models to "think step-by-step," improving their performance on reasoning tasks (math, logic puzzles).
    *   **Pros:** Learn from data, handle fuzziness and variation well, don't always require explicit rules.
    *   **Cons:** Reasoning can be implicit and hard to verify ("black box"), prone to hallucination or logical errors, reasoning might be sophisticated pattern matching rather than true understanding (an ongoing debate).

3.  **Neuro-Symbolic AI:**
    *   A hybrid approach aiming to combine the strengths of symbolic AI (explicit knowledge, robust logic) and neural networks (learning from data, robustness to noise).
    *   This is an active area of research, exploring ways to integrate logic rules with neural network architectures.

**Why are Reasoning Models Important?**

*   **Solving Complex Problems:** Essential for tasks requiring multi-step planning, logical deduction, and understanding intricate relationships (e.g., scientific discovery, mathematical proofs, complex logistics).
*   **Explainable AI (XAI):** Crucial for high-stakes domains (medicine, finance, law) where understanding *why* an AI made a decision is vital.
*   **Robustness and Generalization:** Models that truly reason might be better at handling novel situations not seen during training.
*   **Human-AI Collaboration:** Enables AI systems to understand context, intent, and engage in more meaningful dialogue.

In summary, reasoning models represent a significant step beyond simple pattern recognition in AI, aiming to imbue machines with the capacity for logical thought, problem-solving, and deeper understanding, much like humans.

---

To try out the Gemini 2.5 experimental model:
```
https://us-west1-ttzeng-gcp.cloudfunctions.net/redis-cloud?model={"provider":"gemini","model":"gemini-2.5-pro-exp-03-25","parameters":{}}
```

[1]: https://blog.google/technology/google-deepmind/gemini-model-thinking-updates-march-2025/#gemini-2-5-thinking
